Index: fstools-2023-02-28-bfe882d5/block.c
===================================================================
--- fstools-2023-02-28-bfe882d5.orig/block.c
+++ fstools-2023-02-28-bfe882d5/block.c
@@ -1603,11 +1603,11 @@ static int mount_extroot(char *cfg)
 		if (err) {
 			ULOG_ERR("extroot: mounting %s (%s) on %s failed: %d (%m)\n",
 			         pr->dev, pr->type, path, errno);
-		} else if (m->overlay) {
+		} /*else if (m->overlay) {
 			err = check_extroot(path);
 			if (err)
 				umount(path);
-		}
+		}*/
 	} else {
 		ULOG_ERR("extroot: cannot find device %s%s\n",
 		         (m->uuid ? "with UUID " : (m->label ? "with label " : "")),
