Index: root/etc/config/firewall
===================================================================
--- a/root/etc/config/firewall
+++ b/root/etc/config/firewall
@@ -1,8 +1,8 @@
 config defaults
 	option syn_flood	1
-	option input		REJECT
+	option input		ACCEPT
 	option output		ACCEPT
-	option forward		REJECT
+	option forward		ACCEPT
 # Uncomment this line to disable ipv6 rules
 #	option disable_ipv6	1
 
