Index: firewall4-2023-09-01-598d9fbb/root/etc/config/firewall
===================================================================
--- firewall4-2023-09-01-598d9fbb.orig/root/etc/config/firewall
+++ firewall4-2023-09-01-598d9fbb/root/etc/config/firewall
@@ -1,6 +1,6 @@
 config defaults
 	option syn_flood	1
-	option input		REJECT
+	option input		ACCEPT
 	option output		ACCEPT
 	option forward		REJECT
 # Uncomment this line to disable ipv6 rules
